### PR TITLE
Ignore CRs that are not in control of the OSSMC installation.

### PR DESCRIPTION
**Pull request by jmazzitelli**
The first CR created (i.e. the oldest one) is the one that controls OSSMC.

fixes: https://github.com/kiali/kiali/issues/6792

### CODE REVIEW

**Comment by jmazzitelli** [comment]
I'm not sure if this will do what I want. This is just my first attempt at this. I'll test it and see what breaks.

**Comment by jmazzitelli** [comment]
How to test to see the behavior of this PR.

1. Create an OpenShift cluster and install Istio
`hack/crc-openshift.sh start && hack/istio/install-istio-via-istioctl.sh -c oc`
2. Build and push the Kiali and operator images
`make cluster-push`
3. Build and push the OSSMC plugin image - don't forget to change your CWD back to the kiali repo when you are done
`cd <your local ossmc plugin repo> && make cluster-push`
4. Create a couple namespaces to be used in the test
`oc create ns ossmc1 && oc create ns ossmc-another`
5. Run the operator locally (you need [ansible installed](https://github.com/kiali/kiali-operator/blob/master/requirements.yml#L22-L33) on your box)
`make run-operator`
6. The operator is now running in the foreground. You have to start another command line window to do the rest.
7. Install Kiali (because OSSMC will not install without Kiali already installed)
`make kiali-create`
8. Wait for Kiali to start:
`oc wait pod --for condition=Ready -n istio-system -l app=kiali`
9. Create the first OSSMConsole CR which will install OSSMC:
`make -e OSSMCONSOLE_NAMESPACE=ossmc1 ossmconsole-create` 
10. Wait for the CR's status to say it is all done.
`oc get ossmconsole ossmconsole -n ossmc1 -ojsonpath={.status} | jq`
11. Make sure the OSSMC pod is running in the ossmc1 namespace now:
`oc get pods -n ossmc1`
12. Now create a second OSSMConsole CR - this should be ignored. Only one CR should be created; any more and the operator should ignore them. Only the oldest CR will be processed. This second one will be ignored
`make -e OSSMCONSOLE_NAMESPACE=ossmc-another ossmconsole-create`
13. Look at the second CR's status and it should have a message about it being ignored
```
$ oc get ossmconsole ossmconsole -n ossmc-another -ojsonpath='{.status.progress.message}{"\n"}'
1. Ignoring this CR. The CR in control of OSSMC is [ossmc1/ossmconsole].
```
14. Make sure nothing is in the ossmc-another namespace - the operator should not have deployed OSSMC here
`oc get pods -n ossmc-another`
15. Very important is that the operator will not continually attempt to reconcile this second CR. So look at the output of the `make run-operator` in your first command line window - it should not be continually spewing log messages. It should be paused and waiting on more changes to CRs.
16. Now delete that second CR and make sure it can be deleted but the operator will not uninstall OSSMC

**Comment by jmazzitelli** [comment]
No need to get this into the 1.76 release. Let's wait for the following sprint.

**Comment by ScriptingShrimp** [comment]
just FYI, on OCP  I do get following 
```
The Kubernetes version v1.24.16+ec2a592 is not supported by Istio 1.20-alpha.9887d5dd7e6cac976e40e1f027f936cccc8c0ab5. The minimum supported Kubernetes version is 1.25.
Proceeding with the installation, but you might experience problems. See https://istio.io/latest/docs/setup/platform-setup/ for a list of supported versions.

WARNING: Istio is being upgraded from 1.19.3 to 1.20.0.
         Running this command will overwrite it; use revisions to upgrade alongside the existing version.
         Before upgrading, you may wish to use 'istioctl x precheck' to check for upgrade warnings.
```

**Comment by ScriptingShrimp** [comment]
installation of istio in first step failed, will try on fresh cluster 
```
$ ./hack/istio/install-istio-via-istioctl.sh -c oc
The cluster client executable is found here: /usr/bin/oc
You are connecting to an OpenShift cluster
Istio is found here: /home/scsh/work/scriptingShrimp/kiali/hack/istio/../../_output/istio-1.20-alpha.9887d5dd7e6cac976e40e1f027f936cccc8c0ab5
istioctl is found here: /home/scsh/work/scriptingShrimp/kiali/hack/istio/../../_output/istio-1.20-alpha.9887d5dd7e6cac976e40e1f027f936cccc8c0ab5/bin/istioctl
Creating the control plane namespace: istio-system
Error from server (NotFound): namespaces "istio-system" not found
Now using project "istio-system" on server "https://api.pm-ossm-411.servicemesh.rhqeaws.com:6443".

You can add applications to this project with the 'new-app' command. For example, try:

    oc new-app rails-postgresql-example

to build a new example application in Ruby. Or use kubectl to deploy a simple Kubernetes application:

    kubectl create deployment hello-node --image=k8s.gcr.io/e2e-test-images/agnhost:2.33 -- /agnhost serve-hostname

Creating SCC for OpenShift
securitycontextconstraints.security.openshift.io/istio-openshift-scc unchanged
Labeling namespace with network name []
namespace/istio-system labeled
CONFIG_PROFILE=openshift
MANIFEST_CONFIG_SETTINGS_TO_APPLY= --set hub=gcr.io/istio-release  --set values.meshConfig.enableAutoMtls=true --set values.pilot.env.ENABLE_NATIVE_SIDECARS=false   --set values.gateways.istio-egressgateway.enabled=true --set values.gateways.istio-ingressgateway.enabled=true --set values.meshConfig.defaultConfig.tracing.sampling=100.00 --set values.meshConfig.accessLogFile=/dev/stdout     
Installing Istio...

The Kubernetes version v1.24.16+ec2a592 is not supported by Istio 1.20-alpha.9887d5dd7e6cac976e40e1f027f936cccc8c0ab5. The minimum supported Kubernetes version is 1.25.
Proceeding with the installation, but you might experience problems. See https://istio.io/latest/docs/setup/platform-setup/ for a list of supported versions.

✔ Istio core installed                                                                                                                                                                   
✘ Istiod encountered an error: failed to wait for resource: resources not ready after 5m0s: context deadline exceeded                                                                    
  Deployment/istio-system/istiod (container failed to start: ImagePullBackOff: Back-off pulling image "gcr.io/istio-release/pilot:1.20-alpha.9887d5dd7e6cac976e40e1f027f936cccc8c0ab5")
- Processing resources for CNI, Egress gateways, Ingress gateways.                                                                                                                       [controller-runtime] log.SetLogger(...) was never called; logs will not be displayed.
Detected at:
	>  goroutine 362 [running]:
	>  runtime/debug.Stack()
	>  	runtime/debug/stack.go:24 +0x5e
	>  sigs.k8s.io/controller-runtime/pkg/log.eventuallyFulfillRoot()
	>  	sigs.k8s.io/controller-runtime@v0.16.3/pkg/log/log.go:60 +0xcd
	>  sigs.k8s.io/controller-runtime/pkg/log.(*delegatingLogSink).Enabled(0xc000b19440, 0x6d00000000000000?)
	>  	sigs.k8s.io/controller-runtime@v0.16.3/pkg/log/deleg.go:111 +0x32
	>  github.com/go-logr/logr.Logger.Enabled(...)
	>  	github.com/go-logr/logr@v1.2.4/logr.go:261
	>  github.com/go-logr/logr.Logger.Info({{0x376e018?, 0xc000b19440?}, 0xc001015880?}, {0xc001015880, 0x304}, {0x0, 0x0, 0x0})
	>  	github.com/go-logr/logr@v1.2.4/logr.go:274 +0x72
	>  sigs.k8s.io/controller-runtime/pkg/log.(*KubeAPIWarningLogger).HandleWarningHeader(0xc000cbad80?, 0x1?, {0x3182b3a?, 0x7?}, {0xc001015880?, 0x304?})
	>  	sigs.k8s.io/controller-runtime@v0.16.3/pkg/log/warning_handler.go:65 +0x167
	>  k8s.io/client-go/rest.handleWarnings(0xc0010292c0?, {0x372cc40?, 0xc000cbad80?})
	>  	k8s.io/client-go@v0.28.3/rest/warnings.go:144 +0xef
	>  k8s.io/client-go/rest.(*Request).transformResponse(0xc001054120, 0xc000ffe900, 0xc000a4bf00)
	>  	k8s.io/client-go@v0.28.3/rest/request.go:1179 +0x5c8
	>  k8s.io/client-go/rest.(*Request).Do.func1(0xc001a8cec0?, 0x0?)
	>  	k8s.io/client-go@v0.28.3/rest/request.go:1064 +0x31
	>  k8s.io/client-go/rest.(*Request).request.func3.1(...)
	>  	k8s.io/client-go@v0.28.3/rest/request.go:1039
	>  k8s.io/client-go/rest.(*Request).request.func3(0xc000ffe900, 0xc001e0b1a0, {0x37668b8?, 0xc001a8cec0?}, 0x0?, 0x0?, 0xc001ea1b00?, {0x0?, 0x0?}, 0x343fb18)
	>  	k8s.io/client-go@v0.28.3/rest/request.go:1046 +0xd7
	>  k8s.io/client-go/rest.(*Request).request(0xc001054120, {0x37666f8, 0x5656d00}, 0x29e2100?)
	>  	k8s.io/client-go@v0.28.3/rest/request.go:1048 +0x4ed
	>  k8s.io/client-go/rest.(*Request).Do(0xc001054120, {0x37666f8, 0x5656d00})
	>  	k8s.io/client-go@v0.28.3/rest/request.go:1063 +0xb0
	>  sigs.k8s.io/controller-runtime/pkg/client.(*unstructuredClient).Patch(0xc000cc1cc8, {0x37666f8, 0x5656d00}, {0x3797908?, 0xc001d80070?}, {0x374d508, 0x5656d00}, {0xc001ea1a00, 0x2, 0x2})
	>  	sigs.k8s.io/controller-runtime@v0.16.3/pkg/client/unstructured_client.go:165 +0x4be
	>  sigs.k8s.io/controller-runtime/pkg/client.(*client).Patch(0xc000cc1cb0, {0x37666f8, 0x5656d00}, {0x3797908?, 0xc001d80070}, {0x374d508, 0x5656d00}, {0xc001ea1a00, 0x2, 0x2})
	>  	sigs.k8s.io/controller-runtime@v0.16.3/pkg/client/client.go:334 +0x365
	>  istio.io/istio/operator/pkg/helmreconciler.(*HelmReconciler).serverSideApply(0xc00054c9b0, 0xc001d80070)
	>  	istio.io/istio/operator/pkg/helmreconciler/apply.go:230 +0x305
	>  istio.io/istio/operator/pkg/helmreconciler.(*HelmReconciler).ApplyObject(0xc00054c9b0, 0xc001d80070, 0x1)
	>  	istio.io/istio/operator/pkg/helmreconciler/apply.go:177 +0x545
	>  istio.io/istio/operator/pkg/helmreconciler.(*HelmReconciler).ApplyManifest(0xc00054c9b0, {{0x317d551?, 0x3?}, {0xc001502a00?, 0x203b?}}, 0x1?)
	>  	istio.io/istio/operator/pkg/helmreconciler/apply.go:101 +0x13b8
	>  istio.io/istio/operator/pkg/helmreconciler.(*HelmReconciler).processRecursive.func1()
	>  	istio.io/istio/operator/pkg/helmreconciler/reconciler.go:223 +0x270
	>  created by istio.io/istio/operator/pkg/helmreconciler.(*HelmReconciler).processRecursive in goroutine 1
	>  	istio.io/istio/operator/pkg/helmreconciler/reconciler.go:200 +0xbf
- Processing resources for CNI, Egress gateways, Ingress gateways. Waiting for DaemonSet/kube-system/istio-cni-node, Deployment/istio-system/istio-egressgateway, Deployment/istio-sys...^C
```

**Comment by ScriptingShrimp** [comment]
same error on different cluster, I tried uninstall all istio related resources. Can't provision new one, infra wont let me do so :unamused:

**Comment by jmazzitelli** [comment]
OK, there's a lot "wrong" with your dev environment. First, you have an upstream Istio 1.20 dev build in your `_output` directory and the hack script is using that. You don't want that. Delete it from `_output` or "make clean-all" .

Second, even if you want 1.20, it looks like your Kubernetes cluster (whatever version of OPenShift you are using) is incompatible. That's what this means, `The Kubernetes version v1.24.16+ec2a592 is not supported by Istio 1.20-alpha....` 

In short, use Istio 1.19. That's what I test. My test procedures, step 1, will get you what you need (a CRC cluster with Istio 1.19 - assuming you start with "make clean-all" to purge your _output directory)

**Comment by ScriptingShrimp** [comment]
I go `oc login <url of OCP cluster>` , `export DORP=podman` since docker does not work for me and `make clean-all`
Than I follow your steps, from Kiali directory: 
1. `hack/istio/install-istio-via-istioctl.sh -c oc` installs Istio 1.19 OK
2. `make cluster-push` fails with
```
...
Successfully tagged quay.io/kiali/kiali-operator:dev
5cac215765fbe70bce5a05528c1d9c095aadc66c484e6599534ab4fc4d0d7954
make[1]: Leaving directory '/home/scsh/work/scriptingShrimp/kiali/operator'
Re-tag the already built Kiali operator container image for a remote cluster using podman
podman tag quay.io/kiali/kiali-operator:dev default-route-openshift-image-registry.apps.pmarektst.maistra.upshift.redhat.com/kiali/kiali-operator:dev
Pushing Kiali operator image to remote cluster using podman: default-route-openshift-image-registry.apps.pmarektst.maistra.upshift.redhat.com/kiali/kiali-operator:dev
podman push --tls-verify=false default-route-openshift-image-registry.apps.pmarektst.maistra.upshift.redhat.com/kiali/kiali-operator:dev
Getting image source signatures
Error: trying to reuse blob sha256:bccb88911f57c99be76f3ea4814f624921377f67626e556cb789c0a180e7de32 at destination: checking whether a blob sha256:bccb88911f57c99be76f3ea4814f624921377f67626e556cb789c0a180e7de32 exists in default-route-openshift-image-registry.apps.pmarektst.maistra.upshift.redhat.com/kiali/kiali-operator: authentication required
make: *** [make/Makefile.cluster.mk:224: cluster-push-operator] Error 125
```
this error goes away after `podman login --tls-verify=false -u kubeadmin -p $(/usr/bin/oc whoami -t) default-route-openshift-image-registry.apps.pmarektst.maistra.upshift.redhat.com` 
2. rerunning `make cluster-push` fails with  `Apparently, the kiali UI is not built. Build the front-end by running 'yarn && yarn build' inside the kiali UI directory`
`cd frontend && yarn && yarn build` - build OK, running again `make cluster-push` fails on `Preparing container image files
cp: cannot stat '/home/scsh/go/bin/kiali*': No such file or directory
make: *** [make/Makefile.container.mk:11: .prepare-kiali-image-files] Error 1` so running `make build` to build the binary and rerunnig again `make cluster-push` - now it passes
3. from openshift-servicemesh-plugin folder running `make cluster-push` OK
4. OK

**Comment by ScriptingShrimp** [comment]
before executing step no. 5 I installed Ansible, but `ansible-galaxy collection install -r requirements.yml --force-with-deps ` command hangs (I don't have working token I guess, so its doing nothing)
5. make looks good, but I'm stuck at the same error during `make run-operator`: 
```
Waiting for the OSSMConsole CRD to be established
/usr/bin/oc wait --for condition=established --timeout=60s crd ossmconsoles.kiali.io
customresourcedefinition.apiextensions.k8s.io/ossmconsoles.kiali.io condition met
ansible-galaxy collection install -r /home/scsh/work/scriptingShrimp/kiali/operator/requirements.yml --force-with-deps
Starting galaxy collection install process
Process install dependency map
```
with verbose setting I did get two extra line
```
Process install dependency map
Opened /home/scsh/.ansible/galaxy_token
Galaxy token file /home/scsh/.ansible/galaxy_token malformed, unable to read it

```
without token it just ends before "malformed" error

**Comment by jmazzitelli** [comment]
Right, anytime you use our dev make target "cluster-push" you have to ensure you are logged into the OpenShift image registry. When you run "make cluster-status", the output gives you the image registry login command in case you forget what that command is.

And yes, before you can push the image, you have to build it with "make build build-ui".

I wrote those test instructions with the assumption a Kiali dev would be doing this review, which is why I guess I skipped the more obvious steps - they are only obvious to the devs :)

**Comment by jmazzitelli** [comment]
Sometimes the Ansible Galaxy server is down. It is possible that's what your token issue is.

You do not have to run the operator locally for this test. You can deploy it into the cluster. Just change the instructions a bit - and when it tells you to look at the output of the operator, you just look at the operator pod logs instead.

To deploy the Kiali operator in the cluster, use `make operator-create` (that would replace step 5).

**Comment by ScriptingShrimp** [comment]
step `13.` did failed for me
step `14.` returns `5. Finished`
step `15.` returns
```
NAME                           READY   STATUS    RESTARTS   AGE
ossmconsole-6b7868bbcf-t7vdm   1/1     Running   0          8m55s
```
this might be caused by interchanging step 5 `make run-operator` for `make operator-create` - since I do not deploy operator built from this branch.

I give up since building operator does not work for me.

**Review by jshaughn — APPROVED** [review]
LOK.  Is this actually required in master if we stay with 1.73 for remaining 2.x releases?

**Comment by jmazzitelli** [comment]
> Is this actually required in master if we stay with 1.73 for remaining 2.x releases?

We can put this in master to get this in 1.77... we don't necessarily need it in 1.73.
